src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs

```diff
@@ -35,107 +36,256 @@ public static void DisposeFrames(RenderBatchBuilder batchBuilder, ArrayRange<Ren
+                                // insertions/deletions at smaller sibling indices.
+                                oldKeyItemInfo.OldSiblingIndex = diffContext.SiblingIndex;
+                                keyedItemInfos[oldKey] = oldKeyItemInfo;
+                                newKeyItemInfo.NewSiblingIndex = diffContext.SiblingIndex;
+                                keyedItemInfos[newKey] = newKeyItemInfo;
+                            }
+                            else
+                            {
+                                // It's an insertion or a deletion, or both
+                                // If the new key is in both trees, but the old key isn't, then the old item was deleted
+                                // Otherwise, it's either an insertion or *both* insertion+deletion, so pick insertion and get the deletion on the next iteration if needed
+                                action = newKeyIsInOldTree ? DiffAction.Delete: DiffAction.Insert;
+                            }
+                        }
+                        #endregion
                     }
                     else
                     {
-                        // New sequence looped back but old one didn't
-                        // The old sequence either has some extra trailing elements in the current loop block
-                        // which we should delete, or the new sequence has extra trailing loop blocks which we
-                        // should insert
-                        // TODO: Find a way of not recomputing this next flag on every iteration
-                        var oldLoopsBackLater = false;
-                        for (var testIndex = oldStartIndex + 1; testIndex < oldEndIndexExcl; testIndex++)
+                        #region "Get diff action for non-matching sequence numbers"
+                        // Sequence doesn't match
+                        var oldLoopedBack = oldSeq <= prevOldSeq;
+                        var newLoopedBack = newSeq <= prevNewSeq;
+                        if (oldLoopedBack == newLoopedBack)
                         {
-                            if (oldTree[testIndex].Sequence < oldSeq)
+                            // Both sequences are proceeding through the same loop block, so do a simple
+                            // preordered merge join (picking from whichever side brings us closer to being
+                            // back in sync)
+                            action = newSeq < oldSeq ? DiffAction.Insert : DiffAction.Delete;
+
+                            if (oldLoopedBack)
                             {
-                                oldLoopsBackLater = true;
-                                break;
+                                // If both old and new have now looped back, we must reset their 'looped back'
+                                // tracker so we can treat them as proceeding through the same loop block
+                                prevOldSeq = prevNewSeq = -1;
```

Comment:
nit: We don't do this through the codebase. It's easy to miss when modifying code.

Reply:
OK, will change it.